// @graphql-tools/batch-delegate/es5/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

const tslib = require('tslib');
const graphql = require('graphql');
const DataLoader = _interopDefault(require('dataloader'));
const delegate = require('@graphql-tools/delegate/es5');
const utils = require('@graphql-tools/utils/es5');

function createBatchFn(options) {
    var _a, _b;
    var argsFromKeys = (_a = options.argsFromKeys) !== null && _a !== void 0 ? _a : (function (keys) { return ({ ids: keys }); });
    var fieldName = (_b = options.fieldName) !== null && _b !== void 0 ? _b : options.info.fieldName;
    var valuesFromResults = options.valuesFromResults, lazyOptionsFn = options.lazyOptionsFn;
    return function batchFn(keys) {
        return tslib.__awaiter(this, void 0, void 0, function () {
            var results, values;
            return tslib.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, delegate.delegateToSchema(tslib.__assign({ returnType: new graphql.GraphQLList(graphql.getNamedType(options.info.returnType)), onLocatedError: function (originalError) {
                                if (originalError.path == null) {
                                    return originalError;
                                }
                                var _a = tslib.__read(originalError.path, 2), pathFieldName = _a[0], pathNumber = _a[1];
                                if (pathFieldName !== fieldName) {
                                    return originalError;
                                }
                                var pathNumberType = typeof pathNumber;
                                if (pathNumberType !== 'number') {
                                    return originalError;
                                }
                                return utils.relocatedError(originalError, originalError.path.slice(0, 0).concat(originalError.path.slice(2)));
                            }, args: argsFromKeys(keys) }, (lazyOptionsFn == null ? options : lazyOptionsFn(options))))];
                    case 1:
                        results = _a.sent();
                        if (results instanceof Error) {
                            return [2 /*return*/, keys.map(function () { return results; })];
                        }
                        values = valuesFromResults == null ? results : valuesFromResults(results, keys);
                        return [2 /*return*/, Array.isArray(values) ? values : keys.map(function () { return values; })];
                }
            });
        });
    };
}
function defaultCacheKeyFn(key) {
    if (typeof key === 'object') {
        return JSON.stringify(key);
    }
    return key;
}
var getLoadersMap = utils.memoize2(function getLoadersMap(_fieldNodes, _schema) {
    return new Map();
});
function getLoader(options) {
    var _a;
    var fieldName = (_a = options.fieldName) !== null && _a !== void 0 ? _a : options.info.fieldName;
    var loaders = getLoadersMap(options.info.fieldNodes, options.schema);
    var loader = loaders.get(fieldName);
    // Prevents the keys to be passed with the same structure
    var dataLoaderOptions = tslib.__assign({ cacheKeyFn: defaultCacheKeyFn }, options.dataLoaderOptions);
    if (loader === undefined) {
        var batchFn = createBatchFn(options);
        loader = new DataLoader(batchFn, dataLoaderOptions);
        loaders.set(fieldName, loader);
    }
    return loader;
}

function batchDelegateToSchema(options) {
    var key = options.key;
    if (key == null) {
        return null;
    }
    else if (Array.isArray(key) && !key.length) {
        return [];
    }
    var loader = getLoader(options);
    return Array.isArray(key) ? loader.loadMany(key) : loader.load(key);
}

function createBatchDelegateFn(optionsOrArgsFromKeys, lazyOptionsFn, dataLoaderOptions, valuesFromResults) {
    return typeof optionsOrArgsFromKeys === 'function'
        ? createBatchDelegateFnImpl({
            argsFromKeys: optionsOrArgsFromKeys,
            lazyOptionsFn: lazyOptionsFn,
            dataLoaderOptions: dataLoaderOptions,
            valuesFromResults: valuesFromResults,
        })
        : createBatchDelegateFnImpl(optionsOrArgsFromKeys);
}
function createBatchDelegateFnImpl(options) {
    return function (batchDelegateOptions) {
        var loader = getLoader(tslib.__assign(tslib.__assign({}, options), batchDelegateOptions));
        return loader.load(batchDelegateOptions.key);
    };
}

exports.batchDelegateToSchema = batchDelegateToSchema;
exports.createBatchDelegateFn = createBatchDelegateFn;
